src/NovelRT/Experimental/Graphics/GraphicsMemoryBlockCollection.cpp

```diff
@@ -0,0 +1,481 @@
+
+        if (!canCreateNewBlock)
+        {
+            return false;
+        }
+
+        size_t blockSize = GetAdjustedBlockSize(sizeWithMargins);
+
+        if (blockSize >= availableMemory)
+        {
+            return false;
+        }
+
+        std::shared_ptr<GraphicsMemoryBlock> block = AddBlock(blockSize);
+        return block->TryAllocate(size, alignment, region);
+    }
+
+    GraphicsMemoryBlockCollection::GraphicsMemoryBlockCollection(std::shared_ptr<GraphicsDevice> device,
+                                                                 std::shared_ptr<GraphicsMemoryAllocator> allocator)
+        : GraphicsDeviceObject(std::move(device)),
+          _allocator(std::move(allocator)),
+          _blocks{},
+          _mutex(),
+          _emptyBlock(nullptr),
+          _minimumSize(0),
+          _size(0LL),
+          _state()
+    {
+        if (_allocator == nullptr)
+        {
+            throw Exceptions::NullPointerException("Parameter name: allocator.");
+        }
+
+        if (_allocator->GetDevice() != GraphicsDeviceObject::GetDevice())
+        {
+            throw std::out_of_range(
+                "The same graphics device is not being used for this memory block collection. It is out "
+                "of range of one of the devices.");
+        }
+
+        const GraphicsMemoryAllocatorSettings& settings = _allocator->GetSettings();
+
+        int32_t minimumBlockCount = settings.MinimumBlockCountPerCollection;
+        size_t maximumSharedBlockSize = settings.MaximumSharedBlockSize.value_or(0);
+
+        for (int32_t i = 0; i < minimumBlockCount; ++i)
+        {
+            size_t blockSize = GetAdjustedBlockSize(maximumSharedBlockSize);
+            static_cast<void>(AddBlock(blockSize));
+        }
+
+        static_cast<void>(_state.Transition(Threading::VolatileState::Initialised));
+    }
+
+    GraphicsMemoryRegion<GraphicsMemoryBlock> GraphicsMemoryBlockCollection::Allocate(
+        size_t size,
+        size_t alignment,
+        GraphicsMemoryRegionAllocationFlags flags)
+    {
+        GraphicsMemoryRegion<GraphicsMemoryBlock> outRegion(0, nullptr, nullptr, false, 0, 0);
+        bool succeeded = TryAllocate(size, alignment, flags, outRegion);
+
+        if (!succeeded)
+        {
+            throw Exceptions::OutOfMemoryException("Attempted to allocate memory region of size: " +
+                                                   std::to_string(size));
+        }
+
+        return outRegion;
+    }
+
+    void GraphicsMemoryBlockCollection::Free(const GraphicsMemoryRegion<GraphicsMemoryBlock>& region)
+    {
+        std::shared_ptr<GraphicsMemoryBlock> block = region.GetCollection();
+
+        if (block == nullptr)
+        {
+            throw Exceptions::NullPointerException("The supplied memory region's collection is nullptr.");
+        }
+
+        std::lock_guard<std::mutex> guard(_mutex); // TODO: come back to this when Tanner gets back to you
```

Comment:
Did Tanner ever get back to you on this?

Reply:
Yeah I had a discussion with him, but I think this ultimately boiled down to "leave it alone for now".